# JSON

JavaScript Object Notation의 약자로 사람이 읽을 수 있는 텍스트 기반의 데이터 교환 표준  
JSON은 기존의 데이터를 주고받기 위해 사용하던 XML의 대안으로 좀 더 쉽게 데이터를 교환하고 저장하기 위하여 고안


### 밑의 예제는 공공데이터포털에서 가져온 기상청_동네예보 조회서비스 데이터

In [6]:
import requests
import json

url = 'http://apis.data.go.kr/1360000/VilageFcstInfoService/getUltraSrtNcst'
queryParams = 'ServiceKey=GtbNVjtp2jvReVCmZ3V%2F4mPHVV02KkS3%2B86s%2BApiM4dnx8KoVGUWqmRZj7rH%2FJ96%2Ff9ZkBPMvTBTjNGuLoLaGA%3D%3D&pageNo=1&numOfRows=10&dataType=JSON&base_date=20210726&base_time=0600&nx=18&ny=1'

res = requests.get(url+'?'+queryParams)

result = json.loads(res.content.decode('utf8').replace("'", '"'))

print(result)


{'response': {'header': {'resultCode': '00', 'resultMsg': 'NORMAL_SERVICE'}, 'body': {'dataType': 'JSON', 'items': {'item': [{'baseDate': '20210726', 'baseTime': '0600', 'category': 'PTY', 'nx': 18, 'ny': 1, 'obsrValue': '-999'}, {'baseDate': '20210726', 'baseTime': '0600', 'category': 'REH', 'nx': 18, 'ny': 1, 'obsrValue': '-998'}, {'baseDate': '20210726', 'baseTime': '0600', 'category': 'RN1', 'nx': 18, 'ny': 1, 'obsrValue': '-998.9'}, {'baseDate': '20210726', 'baseTime': '0600', 'category': 'T1H', 'nx': 18, 'ny': 1, 'obsrValue': '-999'}, {'baseDate': '20210726', 'baseTime': '0600', 'category': 'UUU', 'nx': 18, 'ny': 1, 'obsrValue': '-998.9'}, {'baseDate': '20210726', 'baseTime': '0600', 'category': 'VEC', 'nx': 18, 'ny': 1, 'obsrValue': '-998'}, {'baseDate': '20210726', 'baseTime': '0600', 'category': 'VVV', 'nx': 18, 'ny': 1, 'obsrValue': '-998.9'}, {'baseDate': '20210726', 'baseTime': '0600', 'category': 'WSD', 'nx': 18, 'ny': 1, 'obsrValue': '-998.9'}]}, 'pageNo': 1, 'numOfRows':

이 결과 데이터를 복사해서 json beautify 사이트들에 붙여 넣으면  
잘 정리해서 데이터를 보여준다. 거기서 찾고 싶은 데이터를 찾아서 인덱스하자

-------------------------------------------

-------------------------------------------

In [9]:
result['response']['body']['items']['item']

[{'baseDate': '20210726',
  'baseTime': '0600',
  'category': 'PTY',
  'nx': 18,
  'ny': 1,
  'obsrValue': '-999'},
 {'baseDate': '20210726',
  'baseTime': '0600',
  'category': 'REH',
  'nx': 18,
  'ny': 1,
  'obsrValue': '-998'},
 {'baseDate': '20210726',
  'baseTime': '0600',
  'category': 'RN1',
  'nx': 18,
  'ny': 1,
  'obsrValue': '-998.9'},
 {'baseDate': '20210726',
  'baseTime': '0600',
  'category': 'T1H',
  'nx': 18,
  'ny': 1,
  'obsrValue': '-999'},
 {'baseDate': '20210726',
  'baseTime': '0600',
  'category': 'UUU',
  'nx': 18,
  'ny': 1,
  'obsrValue': '-998.9'},
 {'baseDate': '20210726',
  'baseTime': '0600',
  'category': 'VEC',
  'nx': 18,
  'ny': 1,
  'obsrValue': '-998'},
 {'baseDate': '20210726',
  'baseTime': '0600',
  'category': 'VVV',
  'nx': 18,
  'ny': 1,
  'obsrValue': '-998.9'},
 {'baseDate': '20210726',
  'baseTime': '0600',
  'category': 'WSD',
  'nx': 18,
  'ny': 1,
  'obsrValue': '-998.9'}]

In [11]:
for i in result['response']['body']['items']['item'] :
    print(i['category'])

PTY
REH
RN1
T1H
UUU
VEC
VVV
WSD


## 문제

category 값이 VVV인 obsrValue의 값을 출력하세요.  
정답 :-998.9

In [13]:
for i in result['response']['body']['items']['item'] :
    if i['category'] =='VVV' :
        print(i['obsrValue'])

-998.9


In [16]:
import requests
import json

url = 'http://apis.data.go.kr/B552584/ArpltnInforInqireSvc/getMinuDustFrcstDspth'
queryParams = 'ServiceKey=GtbNVjtp2jvReVCmZ3V%2F4mPHVV02KkS3%2B86s%2BApiM4dnx8KoVGUWqmRZj7rH%2FJ96%2Ff9ZkBPMvTBTjNGuLoLaGA%3D%3D&returnType=json&numOfRows=100&pageNo=1&searchDate=2021-07-25&InformCode=PM10'

res = requests.get(url+'?'+queryParams)

result1 = json.loads(res.content.decode('utf8'))

print(result1)


{'response': {'body': {'totalCount': 24, 'items': [{'imageUrl4': 'https://www.airkorea.or.kr/file/proxyImage?fileName=2021/07/25/AQFv1_15h.20210725.KNU_09_01.PM2P5.1hsp.2021072521.png', 'informCode': 'PM10', 'imageUrl5': 'https://www.airkorea.or.kr/file/proxyImage?fileName=2021/07/25/AQFv1_15h.20210725.KNU_09_01.PM2P5.1hsp.2021072603.png', 'imageUrl6': 'https://www.airkorea.or.kr/file/proxyImage?fileName=2021/07/25/AQFv1_15h.20210725.KNU_09_01.PM2P5.1hsp.2021072609.png', 'actionKnack': None, 'informCause': '○ [미세먼지] 원활한 대기 확산으로 대기 상태가 대체로 청정할 것으로 예상됩니다.', 'informOverall': "○ [미세먼지] 전 권역이 '좋음'∼'보통'으로 예상됩니다.", 'informData': '2021-07-25', 'informGrade': '서울 : 좋음,제주 : 좋음,전남 : 좋음,전북 : 좋음,광주 : 좋음,경남 : 좋음,경북 : 좋음,울산 : 좋음,대구 : 좋음,부산 : 좋음,충남 : 좋음,충북 : 좋음,세종 : 좋음,대전 : 좋음,영동 : 좋음,영서 : 좋음,경기남부 : 좋음,경기북부 : 좋음,인천 : 좋음', 'dataTime': '2021-07-25 23시 발표', 'imageUrl3': 'https://www.airkorea.or.kr/file/proxyImage?fileName=2021/07/25/AQFv1_15h.20210725.KNU_09_01.PM10.1hsp.2021072609.png', 'imageUrl2': 'ht

## 문제
informGrade, dataTime 을 출력

In [20]:
for i in result1['response']['body']['items'] :
    print('informGrade : '+i['informGrade'])
    print('dataTime : ' + i['dataTime'])

informGrade : 서울 : 좋음,제주 : 좋음,전남 : 좋음,전북 : 좋음,광주 : 좋음,경남 : 좋음,경북 : 좋음,울산 : 좋음,대구 : 좋음,부산 : 좋음,충남 : 좋음,충북 : 좋음,세종 : 좋음,대전 : 좋음,영동 : 좋음,영서 : 좋음,경기남부 : 좋음,경기북부 : 좋음,인천 : 좋음
dataTime : 2021-07-25 23시 발표
informGrade : 서울 : 좋음,제주 : 좋음,전남 : 좋음,전북 : 좋음,광주 : 좋음,경남 : 좋음,경북 : 좋음,울산 : 좋음,대구 : 좋음,부산 : 좋음,충남 : 좋음,충북 : 좋음,세종 : 좋음,대전 : 좋음,영동 : 좋음,영서 : 좋음,경기남부 : 좋음,경기북부 : 좋음,인천 : 좋음
dataTime : 2021-07-25 23시 발표
informGrade : 서울 : 좋음,제주 : 좋음,전남 : 좋음,전북 : 좋음,광주 : 좋음,경남 : 좋음,경북 : 좋음,울산 : 좋음,대구 : 좋음,부산 : 좋음,충남 : 좋음,충북 : 좋음,세종 : 좋음,대전 : 좋음,영동 : 좋음,영서 : 좋음,경기남부 : 좋음,경기북부 : 좋음,인천 : 좋음
dataTime : 2021-07-25 23시 발표
informGrade : 서울 : 좋음,제주 : 좋음,전남 : 좋음,전북 : 좋음,광주 : 좋음,경남 : 좋음,경북 : 좋음,울산 : 좋음,대구 : 좋음,부산 : 좋음,충남 : 보통,충북 : 좋음,세종 : 좋음,대전 : 좋음,영동 : 좋음,영서 : 좋음,경기남부 : 좋음,경기북부 : 좋음,인천 : 보통
dataTime : 2021-07-25 23시 발표
informGrade : 서울 : 좋음,제주 : 좋음,전남 : 좋음,전북 : 좋음,광주 : 좋음,경남 : 좋음,경북 : 좋음,울산 : 좋음,대구 : 좋음,부산 : 좋음,충남 : 좋음,충북 : 좋음,세종 : 좋음,대전 : 좋음,영동 : 좋음,영서 : 좋음,경기남부 : 좋음,경기북부 : 좋음,인천 : 좋음
dataTime : 2021-07-25 23시 발표
infor

## 연습
2019년 5월 8일 거래량 데이터 불러오기

In [21]:
stock = {
	"2019.05.10": {
		"종가": "57,400",
		"전일비": "400",
		"시가": "57,500",
		"고가": "59,000",
		"저가": "56,300",
		"거래량": "861,478"
	},
	"2019.05.09": {
		"종가": "57,800",
		"전일비": "3,200",
		"시가": "60,900",
		"고가": "61,500",
		"저가": "57,700",
		"거래량": "1,237,743"
	},
	"2019.05.08": {
		"종가": "61,000",
		"전일비": "1,700",
		"시가": "62,000",
		"고가": "65,200",
		"저가": "60,300",
		"거래량": "1,625,296"
	},
	"2019.05.07": {
		"종가": "62,700",
		"전일비": "1,100",
		"시가": "63,000",
		"고가": "63,500",
		"저가": "62,700",
		"거래량": "414,906"
	}
}

In [22]:
stock['2019.05.08']['거래량']

'1,625,296'

## 문제
각 날짜별로 시가와 종가를 출력하세요.
  
  날짜  
  시가  
  종가   >> 이르케

In [25]:
for i,j in stock.items() :
    print(i)
    print(j['시가'])
    print(j['종가'])
    print(' ')

2019.05.10
57,500
57,400
 
2019.05.09
60,900
57,800
 
2019.05.08
62,000
61,000
 
2019.05.07
63,000
62,700
 
